// popmotion/lib/popmotion.js
'use strict';

exports.__esModule = true;
exports.svgPath = exports.svg = exports.css = exports.Renderer = exports.value = exports.spring = exports.stagger = exports.tween = exports.trackOffset = exports.touches = exports.pointer = exports.physics = exports.parallel = exports.delay = exports.crossFade = exports.composite = exports.colorTween = exports.chain = exports.Action = exports.valueTypes = exports.transform = exports.easing = exports.calc = exports.currentFrameTimestamp = exports.timeSinceLastFrame = exports.cancelOnFrameEnd = exports.cancelOnFrameRender = exports.cancelOnFrameUpdate = exports.cancelOnFrameStart = exports.onFrameEnd = exports.onFrameRender = exports.onFrameUpdate = exports.onFrameStart = undefined;

var _framesync = require('framesync');

Object.defineProperty(exports, 'onFrameStart', {
  enumerable: true,
  get: function get() {
    return _framesync.onFrameStart;
  }
});
Object.defineProperty(exports, 'onFrameUpdate', {
  enumerable: true,
  get: function get() {
    return _framesync.onFrameUpdate;
  }
});
Object.defineProperty(exports, 'onFrameRender', {
  enumerable: true,
  get: function get() {
    return _framesync.onFrameRender;
  }
});
Object.defineProperty(exports, 'onFrameEnd', {
  enumerable: true,
  get: function get() {
    return _framesync.onFrameEnd;
  }
});
Object.defineProperty(exports, 'cancelOnFrameStart', {
  enumerable: true,
  get: function get() {
    return _framesync.cancelOnFrameStart;
  }
});
Object.defineProperty(exports, 'cancelOnFrameUpdate', {
  enumerable: true,
  get: function get() {
    return _framesync.cancelOnFrameUpdate;
  }
});
Object.defineProperty(exports, 'cancelOnFrameRender', {
  enumerable: true,
  get: function get() {
    return _framesync.cancelOnFrameRender;
  }
});
Object.defineProperty(exports, 'cancelOnFrameEnd', {
  enumerable: true,
  get: function get() {
    return _framesync.cancelOnFrameEnd;
  }
});
Object.defineProperty(exports, 'timeSinceLastFrame', {
  enumerable: true,
  get: function get() {
    return _framesync.timeSinceLastFrame;
  }
});
Object.defineProperty(exports, 'currentFrameTimestamp', {
  enumerable: true,
  get: function get() {
    return _framesync.currentFrameTime;
  }
});

var _calc2 = require('./inc/calc');

var _calc = _interopRequireWildcard(_calc2);

var _easing2 = require('./inc/easing');

var _easing = _interopRequireWildcard(_easing2);

var _transformers = require('./inc/transformers');

var _transform = _interopRequireWildcard(_transformers);

var _valueTypes2 = require('./inc/value-types');

var _valueTypes = _interopRequireWildcard(_valueTypes2);

var _actions = require('./actions');

var _actions2 = _interopRequireDefault(_actions);

var _chain2 = require('./actions/chain');

var _chain3 = _interopRequireDefault(_chain2);

var _colorTween2 = require('./actions/color-tween');

var _colorTween3 = _interopRequireDefault(_colorTween2);

var _composite2 = require('./actions/composite');

var _composite3 = _interopRequireDefault(_composite2);

var _crossFade2 = require('./actions/cross-fade');

var _crossFade3 = _interopRequireDefault(_crossFade2);

var _delay2 = require('./actions/delay');

var _delay3 = _interopRequireDefault(_delay2);

var _parallel2 = require('./actions/parallel');

var _parallel3 = _interopRequireDefault(_parallel2);

var _physics2 = require('./actions/physics');

var _physics3 = _interopRequireDefault(_physics2);

var _pointer2 = require('./actions/pointer');

var _pointer3 = _interopRequireDefault(_pointer2);

var _touches2 = require('./actions/touches');

var _touches3 = _interopRequireDefault(_touches2);

var _trackOffset2 = require('./actions/track-offset');

var _trackOffset3 = _interopRequireDefault(_trackOffset2);

var _tween2 = require('./actions/tween');

var _tween3 = _interopRequireDefault(_tween2);

var _stagger2 = require('./actions/stagger');

var _stagger3 = _interopRequireDefault(_stagger2);

var _spring2 = require('./actions/spring');

var _spring3 = _interopRequireDefault(_spring2);

var _value2 = require('./actions/value');

var _value3 = _interopRequireDefault(_value2);

var _renderers = require('./renderers');

var _renderers2 = _interopRequireDefault(_renderers);

var _css2 = require('./renderers/css');

var _css3 = _interopRequireDefault(_css2);

var _svg2 = require('./renderers/svg');

var _svg3 = _interopRequireDefault(_svg2);

var _svgPath2 = require('./renderers/svg-path');

var _svgPath3 = _interopRequireDefault(_svgPath2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

exports.calc = _calc;
exports.easing = _easing;
exports.transform = _transform;
exports.valueTypes = _valueTypes;

// Actions

exports.Action = _actions2.default;
exports.chain = _chain3.default;
exports.colorTween = _colorTween3.default;
exports.composite = _composite3.default;
exports.crossFade = _crossFade3.default;
exports.delay = _delay3.default;
//export inertia from './actions/inertia';

exports.parallel = _parallel3.default;
exports.physics = _physics3.default;
exports.pointer = _pointer3.default;
exports.touches = _touches3.default;
exports.trackOffset = _trackOffset3.default;
exports.tween = _tween3.default;
exports.stagger = _stagger3.default;
exports.spring = _spring3.default;
exports.value = _value3.default;

// Renderers

exports.Renderer = _renderers2.default;
exports.css = _css3.default;
exports.svg = _svg3.default;
exports.svgPath = _svgPath3.default;
